safe_control_gym/safety_filters/cbf/cbf.py

```diff
@@ -0,0 +1,327 @@
+        assert self.is_control_affine()
+
+        # Lie derivative with respect to the dynamics
+        self.lie_derivative = self.get_lie_derivative()
+
+        self.linear_func = linear_function(self.slope)
+
+        # Setup Optimizer
+        self.setup_optimizer()
+
+        self.reset()
+
+    def get_lie_derivative(self):
+        '''Determines the Lie derivative of the CBF with respect to the known dynamics. '''
+        dVdx = cs.gradient(self.cbf(X=self.X)['cbf'], self.X)
+        LfV = cs.dot(dVdx, self.model.x_dot)
+        LfV_func = cs.Function('LfV', [self.X, self.u], [LfV], ['X', 'u'], ['LfV'])
+        return LfV_func
+
+    def is_control_affine(self):
+        '''Check if the system is control affine. '''
+        dfdu = cs.jacobian(self.model.x_dot, self.u)
+        return not cs.depends_on(dfdu, self.u)
+
+    def setup_optimizer(self):
+        '''Setup the certifying CBF problem. '''
+        nx, nu = self.model.nx, self.model.nu
+
+        # Define optimizer
+        opti = cs.Opti('conic')  # Tell casadi that it's a conic problem
+
+        # Optimization variable: control input
+        u_var = opti.variable(nu, 1)
+
+        # States
+        current_state_var = opti.parameter(nx, 1)
+        uncertified_action_var = opti.parameter(nu, 1)
+
+        # Evaluate at Lie derivative and CBF at the current state
+        lie_derivative_at_x = self.lie_derivative(X=current_state_var, u=u_var)['LfV']
+        barrier_at_x = self.cbf(X=current_state_var)['cbf']
+
+        right_hand_side = 0.0
+        slack_var = opti.variable(1, 1)
+
+        if self.soft_constrained:
+            # Quadratic objective
+            cost = 0.5 * cs.norm_2(uncertified_action_var - u_var) ** 2 + self.slack_weight * slack_var**2
+
+            # Soften CBF constraint
+            right_hand_side = slack_var
+
+            # Non-negativity constraint on slack variable
+            opti.subject_to(slack_var >= 0.0)
+        else:
+            opti.subject_to(slack_var == 0.0)
+            # Quadratic objective
+            cost = 0.5 * cs.norm_2(uncertified_action_var - u_var) ** 2
+
+        # CBF constraint
+        opti.subject_to(-self.linear_func(x=barrier_at_x)['y'] - lie_derivative_at_x <=
+                        right_hand_side)
+
+        # Input constraints
+        for input_constraint in self.input_constraints_sym:
+            opti.subject_to(input_constraint(u_var) <= 0)
+
+        opti.minimize(cost)
+
+        # Set verbosity option of optimizer
+        opts = {'printLevel': 'low', 'error_on_fail': False}
+
+        # Select QP solver
+        opti.solver('qpoases', opts)
+
+        self.opti_dict = {
+            'opti': opti,
+            'u_var': u_var,
+            'slack_var': slack_var,
+            'current_state_var': current_state_var,
+            'uncertified_action_var': uncertified_action_var,
+            'cost': cost
+        }
+
+    def solve_optimization(self,
+                           current_state,
```

Comment:
Add types? `current_state: np.ndarray` etc.

Reply:
We haven't been making types explicit throughout the gym except for in the __init__ function really. I think this is a more general style discussion.

Reply:
@JacopoPan @Justin-Yuan idk if you want to move towards being more explicit on typing

Reply:
Type (and add defaults) wherever possible but I would NOT to make mandatory repo-wide